From patchwork Wed Jul 20 02:17:46 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12923229
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 5EFD1CCA47F
	for <linux-mm@archiver.kernel.org>; Wed, 20 Jul 2022 02:17:59 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id AAE0A6B0078; Tue, 19 Jul 2022 22:17:53 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 9E7906B007B; Tue, 19 Jul 2022 22:17:53 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 7C363940008; Tue, 19 Jul 2022 22:17:53 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0013.hostedemail.com
 [216.40.44.13])
	by kanga.kvack.org (Postfix) with ESMTP id 67AE86B0078
	for <linux-mm@kvack.org>; Tue, 19 Jul 2022 22:17:53 -0400 (EDT)
Received: from smtpin31.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay10.hostedemail.com (Postfix) with ESMTP id 45D2AC036A
	for <linux-mm@kvack.org>; Wed, 20 Jul 2022 02:17:53 +0000 (UTC)
X-FDA: 79705867626.31.A58866B
Received: from mx0b-00069f02.pphosted.com (mx0b-00069f02.pphosted.com
 [205.220.177.32])
	by imf17.hostedemail.com (Postfix) with ESMTP id BF0A440093
	for <linux-mm@kvack.org>; Wed, 20 Jul 2022 02:17:52 +0000 (UTC)
Received: from pps.filterd (m0246632.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 26K1HGGj003844;
	Wed, 20 Jul 2022 02:17:51 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=sbBpb33nkm+Z7jbG5A4u98Fzt0q/0a3alrdyha2Zx9g=;
 b=Sv6GRdkK6WmivZnIh8qfRrFyuQfLhWs7g5awm3892rGirouELFTdv3GtIF3KhhMKSVbZ
 9kDkoZC+5B6KMhKiQ3v2zXn50z1+woEMDWT5XGNHV84sqL8xlBICGZDUMXlbUZJQoDLV
 a9qQJqu2Zq47X1JGpVnZnQyDPN8jw51Zc7AJ/GOxNyVubzKlLWHBLlwvCYF8VFI6ItSl
 G/J7uF8/ZKeceUEHI+xer4veX4D3uWX8cAblc57S5+n4ltFkRsr62I+AXkwz/ZymBaMe
 z2CiHqPEkMwr7UxKsVvXn9NFPVgHFDIoaDGwy6Rlqq2fZIAclcfEB5mc/MK8j6Amxpx7 lA==
Received: from iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com
 (iadpaimrmta03.appoci.oracle.com [130.35.103.27])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3hbmxs83k7-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Wed, 20 Jul 2022 02:17:50 +0000
Received: from pps.filterd
 (iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com [127.0.0.1])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 26JNHLKt022321;
	Wed, 20 Jul 2022 02:17:50 GMT
Received: from nam10-mw2-obe.outbound.protection.outlook.com
 (mail-mw2nam10lp2109.outbound.protection.outlook.com [104.47.55.109])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (PPS) with ESMTPS id
 3hc1hseann-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Wed, 20 Jul 2022 02:17:50 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=ccSpyAGqwoVoVaQJTFu6uBb1crnMQjzJDeDnCbd9je6/w2kdZIneSGg7cB96jUAGg3Ob6P1cgVi/AvyOupmG+hX2vugTx9WFMzPd6x9XyCF7mjyBVvsSCB3P5Cvn0AFl7GaVlnud6NSoWcMd5FGnsoq+lj2UnHts3B3IYMmYJsIFX46SiagG4qy+1cCaF6laYiLmJhl+uu0w1hY3nT5U28Vg+A+uxd2eobfJHjPuxwDZzJ5y6O8X17NIPWyuOpjaCr/+JnyYMB+ptFFOP+ZT5CNjDJkiOy27TF37J2c04X9VwLzXCLSKHlyJGBpfVc1WKB2PvYCekRCLNhmhUCnk9Q==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=sbBpb33nkm+Z7jbG5A4u98Fzt0q/0a3alrdyha2Zx9g=;
 b=cRSytdHYJi6KW7H4Xkk44PJ+WyBNp2ym9PnSeJn7QyTUqVQjk6khQ69msbgNe4C/wWzcq9iTaIwqeahpnnShS5O6dl7pLdRCxuC/+dqHnX+KivWDU43uyzXScs7sIrRsUyywQXflD3Pqw6L8AAkFCPsOq+C89QYzRBG36qZ4I9DJL8giywVEiQV+Cc8LSQQ89arG6EZjlEUCKoikambTrnsvG+KrPr2rbEnJTmq5KUH5sZYJOS8ksuzzczGYINueIJprOUDKal0S3P3sE+s2D/h/YGcSml14/osyMHl1k/uc8l3JaBMa5lxqOXgjoE9/3wrJR0U3VNu0FO33ex/u/g==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=sbBpb33nkm+Z7jbG5A4u98Fzt0q/0a3alrdyha2Zx9g=;
 b=PNHWMU3bJLzXBLznb2+x889JAzXvKf5oNQ2Cg3rrUecXTTfcXv4ELVAtAt/c0rvwlqzklEBVhM/lPzCFlSH1AdDQHEkDkOYG4BDG1TCV/4y+mG92eFlQIVSl5LwwEMpB0mZ7YHy2ajQEvHhXhQNzWm0Sjo/LWfpx+H/jv3GzaX8=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by IA1PR10MB6243.namprd10.prod.outlook.com (2603:10b6:208:3a1::15) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5417.26; Wed, 20 Jul
 2022 02:17:48 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::c4d1:edc3:7d21:7c68]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::c4d1:edc3:7d21:7c68%6]) with mapi id 15.20.5438.024; Wed, 20 Jul 2022
 02:17:47 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>,
        Hugh Dickins <hughd@google.com>
CC: Yu Zhao <yuzhao@google.com>
Subject: [PATCH v12 09/69] mm: add VMA iterator
Thread-Topic: [PATCH v12 09/69] mm: add VMA iterator
Thread-Index: AQHYm97mHRqcmaIFyUaKvHPC4PTamQ==
Date: Wed, 20 Jul 2022 02:17:46 +0000
Message-ID: <20220720021727.17018-10-Liam.Howlett@oracle.com>
References: <20220720021727.17018-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220720021727.17018-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 5c184c50-b87e-42f2-4c4b-08da69f60a1c
x-ms-traffictypediagnostic: IA1PR10MB6243:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 F+oFDPdyDRnN9BCYH64kiJeZyggSWq0x3d61wbyOByf5rOXsL4ueTFLzcpZb2s+chBN7PXHKpnMwRY8ose9Q4FBjRkrf4LIHbDCXnpCqDYWn2tHAOeZaGVanWYBpRQcUUFRA/FJkg8ATMb8QA8BB1FVqWkx3dp8sGy4AiSJElQ95+wkzIC5+L3a92RPQfW6ZK7kX0U0wRnceJKljXHzL87kfMvMJpJ9ySPYsg+PE6LY4nUEV9ttSQYhEEu5nQiPOL//QRSAQM3QDOSgMLfer8m0xbAS3F4iSc4hHRemSDv2VXuySoZD6+NBfAQQ0jo2R3gNWtNpchK7GeSpAcy9sjIl6YhckBJ1ziVdN1r2RpjM8YIZsFgOXKyhM6BAaaBHBO39PHccHnb1e2X+eRHMI1oAaYm6mzhkG/H3KrFBibNhCIbttmSDQtHTlOnAjadRRVXEUzPBO8u/RYgA4ozS4wYNwtzxVr/g0WXn4hcHSsty2O0NaSqACccH4Bw8R6tn7JLiy7IvTauhtJWhqNAQ4rZIKzsbeKQFudRkJuK+V1EbD+kQsWWvkdsf65UPsv4ZP+vzH4/eghN8SuBzlUtgV+lMrjHsI2LKO159yTp/JEOlKXSqMhNUpYpZDWibZMUzKuKfyWy6uc8sX3CuRDyTq8lDVut1Q3gEEHVKo59e46AJ1pBF4cMaCFQkGj3zBV2VIYxXYyEeYz1OtpAWXW0v6wHhhhsK8guTedjj4LmhjM+WH8lT0cBJmtCeduBag86dLCqL9lnu5PD174lXYShlhEyL9gzJ2+D1LiZecoNR0QhwUFwEio0NxKl3vU+TdAwC9SaXXgG38YRoCsXw621YBB0VGiCnlzKaQpH5SDJKdfaeGsusIqmJ5TEn2cMKLHAMH
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(396003)(376002)(39860400002)(366004)(136003)(346002)(2906002)(26005)(1076003)(6486002)(91956017)(8676002)(71200400001)(86362001)(8936002)(5660300002)(38100700002)(83380400001)(66476007)(6512007)(36756003)(122000001)(966005)(38070700005)(316002)(478600001)(44832011)(4326008)(64756008)(66946007)(76116006)(186003)(66556008)(110136005)(66446008)(6506007)(41300700001)(2616005);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?hyK9DqjRVbopU9EufngUN7b?=
	=?iso-8859-1?q?kqA5fjuilPT158WcZH3edL2pg9b6eWaFaIGypLGqM2ZE1qKcjc+nFnh0atN1?=
	=?iso-8859-1?q?0/1zZmfRgFBSsZzbnnJ63yvB5j1mU8on0+QUUwvWEuml+waNotQqsR3Gl52W?=
	=?iso-8859-1?q?OOAISuhZcYOOUMhNLzH/lThdn8b1bkLFNmIxpEVYta4tYuTRcyTKjP1LAOQ7?=
	=?iso-8859-1?q?bbGSCykt7h6Zh99JDwUm05g3wKNsTuxMxLvqCR8K94SpNlc9ilVDagVTbhIp?=
	=?iso-8859-1?q?d3rkLymnImxqIpEtG6NPe2FKQKvygjxj4hOCWXAy32ESM1wGlCyvu2pXWN40?=
	=?iso-8859-1?q?nTcz/hx0gpcxsZC4rrhw8C56XNqfNSr0Rz7BwSEAvuqiddbXh48iREngwd+A?=
	=?iso-8859-1?q?hHDK2THlu5f8rXpVDG+gI+K5QV59fQutBL3+jnNjKphU6WuchspnMM2NJzSy?=
	=?iso-8859-1?q?1UfuWnfphLpIZsH5bYFYemvGiHaD7ab2kMsssAHUP763XTEcUQev8tCNFiQx?=
	=?iso-8859-1?q?JMdRZ7j4cqkZ4KGkVDYbhAbpZxGTvaeOPvorVOFhLHEs9BWczmd8dRa8wusn?=
	=?iso-8859-1?q?v48M0L6teck25H/kynMuMcGQVOvjhx5dMaZ+echhnf2vVB6aeyCInfWAd6p4?=
	=?iso-8859-1?q?60x6N8Ca8dTUZCkPDseBYKzgx5VO3TRBdJv98WCZflP1nlCERlA5dDR3b9/B?=
	=?iso-8859-1?q?KN+0oxEt8gJR3WL2QtGzLcZ7r9OnswiOsn0+mDmc8N8bKiiKanNGPfPL2/53?=
	=?iso-8859-1?q?4/OcqD8Ir4vykVoS9MdOg4kU2nTf18utxnWTXJOFNvVDJyL6NlKKoHkNxjKE?=
	=?iso-8859-1?q?gRSmydCC2wWoA0akd7XNpkbGCJzcYSy6CNo5ECbhceX1T7GMsRyaclNyHt/H?=
	=?iso-8859-1?q?4aSpd//6Rnpluolqtq5mdbDYAwsoL1hZUdpsaHYoamOVK31SZXhOBNwF9Ft7?=
	=?iso-8859-1?q?O8jPxCTnueu+r62gM/pnNllv8Q2pXm+TSx6iFsvpGHDpkzJWqBWPXVykRxUH?=
	=?iso-8859-1?q?zYsv/2w2XU+4rJ1Gi5/AbeX3yHeMDSHMsANR+hQJOorterROL6Jrk5Djt8Wu?=
	=?iso-8859-1?q?lEj4pwSPGlGGjE/WLTdqUlSRUpmwW9FJBm2AxmDIGpz/aTdFVXvE+S1QPffH?=
	=?iso-8859-1?q?YSRpGgmQIKicUjZ1Z5HkVFJXGikFrwnkXjaWMDGbaqYf7HdvOiOwLYPAQWoe?=
	=?iso-8859-1?q?lak0QoHS8dweOa2jcHTobGVMsij75h59lS0dEJucVQRJrLshmB3mzv9+oDpZ?=
	=?iso-8859-1?q?T4tfRRE+986MKS353kFCl7e9QgpZRuU1VG4LMeJdvoy5OlSqsBC+JB+p0Woi?=
	=?iso-8859-1?q?bW3M3sFWWrPUO2IJGMi9axiikddh6GHM6YCubsz2uUxzR9GOEUmUC5ykO/B6?=
	=?iso-8859-1?q?oU25m9BfbgNfTpj1BLjqCOoAO5WSHVIVTMLf89m7xFGNk+Uc2VUhYrTBNr/7?=
	=?iso-8859-1?q?PGU/rdJYkj+jy0mY2j2YGZPz6808Zo1Xf4rFpi1pGbvhRZHUnMyHRiDiZL2H?=
	=?iso-8859-1?q?moCtW8laNpsW8J3FhnwpO8DFcJmazkEZAF2eA5twU3LAlKo7AT1y3/tgoMmN?=
	=?iso-8859-1?q?rncLTDWUn8lwWdFCgsffKH/FsGh9tHg5a1t1Oxu1fb18sb2eTdTkbVRa+WKY?=
	=?iso-8859-1?q?7pAanWz+l0S/gYxlzTSbtyKrvFBf4jnuBYm5aUw=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 5c184c50-b87e-42f2-4c4b-08da69f60a1c
X-MS-Exchange-CrossTenant-originalarrivaltime: 20 Jul 2022 02:17:46.4627
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 4f0xB9hdWLeVP6QgfU72OBkSRXJbu5++1SMH28NETqx8DB6dGbLCLxyfTgsKetPPrVWsKKNbI9fj4SIDTKk1Vw==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: IA1PR10MB6243
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.883,Hydra:6.0.517,FMLib:17.11.122.1
 definitions=2022-07-19_10,2022-07-19_01,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 spamscore=0
 bulkscore=0 malwarescore=0
 mlxscore=0 phishscore=0 mlxlogscore=999 suspectscore=0 adultscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2206140000
 definitions=main-2207200008
X-Proofpoint-GUID: Enf8b4KCiK2p8_Y5RMQLQA2BXreABS-2
X-Proofpoint-ORIG-GUID: Enf8b4KCiK2p8_Y5RMQLQA2BXreABS-2
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1658283472;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=sbBpb33nkm+Z7jbG5A4u98Fzt0q/0a3alrdyha2Zx9g=;
	b=6spwVvmuzF0hRaDW1Q6ZWizeQ6HW+t0CBIFbZ7ih7Urmm/MXUhswL4AXhsXC7gfH8VrbTR
	E2nlGF0DJp5/MTfoomF3J+v9rHGkAokZr+Pw/Mo7ZwFZzfNsEnGQyen7s6UmwmGaz7hQbb
	VVtlZ0J2LOBpkPQC0PsEPW8aONlZUHM=
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1658283472; a=rsa-sha256;
	cv=pass;
	b=bfvGo77SHA1eVSnTzf1jxO+Z0o8OCUbff6HUBnC/V+jbWJ6gliZmdnq+0T+UFuk4U80lKY
	chv9JAb1gwy6A9fHg5CzUWuBS6dSl9UzxRx7F5pEUa9cuQSWiBM0R6k9hCGKRrc7j1YTuY
	U3cC/j2EZu4S0oy7pu+3nXKBcVbV4/8=
ARC-Authentication-Results: i=2;
	imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=Sv6GRdkK;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=PNHWMU3b;
	dmarc=pass (policy=none) header.from=oracle.com;
	spf=none (imf17.hostedemail.com: domain of liam.howlett@oracle.com has no SPF
 policy when checking 205.220.177.32) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-Rspamd-Queue-Id: BF0A440093
Authentication-Results: imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=Sv6GRdkK;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=PNHWMU3b;
	dmarc=pass (policy=none) header.from=oracle.com;
	spf=none (imf17.hostedemail.com: domain of liam.howlett@oracle.com has no SPF
 policy when checking 205.220.177.32) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-Rspamd-Server: rspam12
X-Rspam-User: 
X-Stat-Signature: p749t8j5wpayo4u71ynsuus63bo7tfr5
X-HE-Tag: 1658283472-173816
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Matthew Wilcox (Oracle)" <willy@infradead.org>

This thin layer of abstraction over the maple tree state is for iterating
over VMAs.  You can go forwards, go backwards or ask where the iterator
is.  Rename the existing vma_next() to __vma_next() -- it will be removed
by the end of this series.

Link: https://lkml.kernel.org/r/20220504010716.661115-11-Liam.Howlett@oracle.com
Link: https://lkml.kernel.org/r/20220621204632.3370049-10-Liam.Howlett@oracle.com
Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz>
Reviewed-by: David Hildenbrand <david@redhat.com>
Cc: Catalin Marinas <catalin.marinas@arm.com>
Cc: David Howells <dhowells@redhat.com>
Cc: SeongJae Park <sj@kernel.org>
Cc: Will Deacon <will@kernel.org>
Cc: Davidlohr Bueso <dave@stgolabs.net>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---
 include/linux/mm.h       | 32 ++++++++++++++++++++++++++++++++
 include/linux/mm_types.h | 21 +++++++++++++++++++++
 mm/mmap.c                | 10 +++++-----
 3 files changed, 58 insertions(+), 5 deletions(-)

diff --git a/include/linux/mm.h b/include/linux/mm.h
index 111111111..111111111 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -657,6 +657,38 @@ static inline bool vma_is_accessible(struct vm_area_struct *vma)
 	return vma->vm_flags & VM_ACCESS_FLAGS;
 }
 
+static inline
+struct vm_area_struct *vma_find(struct vma_iterator *vmi, unsigned long max)
+{
+	return mas_find(&vmi->mas, max);
+}
+
+static inline struct vm_area_struct *vma_next(struct vma_iterator *vmi)
+{
+	/*
+	 * Uses vma_find() to get the first VMA when the iterator starts.
+	 * Calling mas_next() could skip the first entry.
+	 */
+	return vma_find(vmi, ULONG_MAX);
+}
+
+static inline struct vm_area_struct *vma_prev(struct vma_iterator *vmi)
+{
+	return mas_prev(&vmi->mas, 0);
+}
+
+static inline unsigned long vma_iter_addr(struct vma_iterator *vmi)
+{
+	return vmi->mas.index;
+}
+
+#define for_each_vma(__vmi, __vma)					\
+	while (((__vma) = vma_next(&(__vmi))) != NULL)
+
+/* The MM code likes to work with exclusive end addresses */
+#define for_each_vma_range(__vmi, __vma, __end)				\
+	while (((__vma) = vma_find(&(__vmi), (__end) - 1)) != NULL)
+
 #ifdef CONFIG_SHMEM
 /*
  * The vma_is_shmem is not inline because it is used only by slow
diff --git a/include/linux/mm_types.h b/include/linux/mm_types.h
index 111111111..111111111 100644
--- a/include/linux/mm_types.h
+++ b/include/linux/mm_types.h
@@ -778,6 +778,27 @@ static inline void lru_gen_use_mm(struct mm_struct *mm)
 
 #endif /* CONFIG_LRU_GEN */
 
+struct vma_iterator {
+	struct ma_state mas;
+};
+
+#define VMA_ITERATOR(name, __mm, __addr)				\
+	struct vma_iterator name = {					\
+		.mas = {						\
+			.tree = &(__mm)->mm_mt,				\
+			.index = __addr,				\
+			.node = MAS_START,				\
+		},							\
+	}
+
+static inline void vma_iter_init(struct vma_iterator *vmi,
+		struct mm_struct *mm, unsigned long addr)
+{
+	vmi->mas.tree = &mm->mm_mt;
+	vmi->mas.index = addr;
+	vmi->mas.node = MAS_START;
+}
+
 struct mmu_gather;
 extern void tlb_gather_mmu(struct mmu_gather *tlb, struct mm_struct *mm);
 extern void tlb_gather_mmu_fullmm(struct mmu_gather *tlb, struct mm_struct *mm);
diff --git a/mm/mmap.c b/mm/mmap.c
index 111111111..111111111 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -633,7 +633,7 @@ static int find_vma_links(struct mm_struct *mm, unsigned long addr,
 }
 
 /*
- * vma_next() - Get the next VMA.
+ * __vma_next() - Get the next VMA.
  * @mm: The mm_struct.
  * @vma: The current vma.
  *
@@ -641,7 +641,7 @@ static int find_vma_links(struct mm_struct *mm, unsigned long addr,
  *
  * Returns: The next VMA after @vma.
  */
-static inline struct vm_area_struct *vma_next(struct mm_struct *mm,
+static inline struct vm_area_struct *__vma_next(struct mm_struct *mm,
 					 struct vm_area_struct *vma)
 {
 	if (!vma)
@@ -1338,7 +1338,7 @@ struct vm_area_struct *vma_merge(struct mm_struct *mm,
 	if (vm_flags & VM_SPECIAL)
 		return NULL;
 
-	next = vma_next(mm, prev);
+	next = __vma_next(mm, prev);
 	area = next;
 	if (area && area->vm_end == end)		/* cases 6, 7, 8 */
 		next = next->vm_next;
@@ -2886,7 +2886,7 @@ static void unmap_region(struct mm_struct *mm,
 		struct vm_area_struct *vma, struct vm_area_struct *prev,
 		unsigned long start, unsigned long end)
 {
-	struct vm_area_struct *next = vma_next(mm, prev);
+	struct vm_area_struct *next = __vma_next(mm, prev);
 	struct mmu_gather tlb;
 
 	lru_add_drain();
@@ -3094,7 +3094,7 @@ int __do_munmap(struct mm_struct *mm, unsigned long start, size_t len,
 		if (error)
 			goto split_failed;
 	}
-	vma = vma_next(mm, prev);
+	vma = __vma_next(mm, prev);
 
 	if (unlikely(uf)) {
 		/*

